### PR TITLE
adding adversarial weight perturbation protocol

### LINKED ISSUE #2164: Adversarial Weight Perturbation based adversarial Training
**Is your feature request related to a problem? Please describe.**
I want to implement an adversarial training approach "Adversarial Weight Perturbation (AWP)" because it is useful for ART in providing a better robustness against evasion attacks and mitigates robust overfitting.

**Describe the solution you'd like**
AWP is an important adversarial training approach because it provides better robustness against adversarial attacks and mitigates robust overfitting. AWP has been proposed in paper "Adversarial Weight Perturbation Helps
Robust Generalization".

Paper link: https://proceedings.neurips.cc/paper/2020/file/1ef91c212e30e14bf125e9374262401f-Paper.pdf

It is also a base component of more advanced adversarial training approaches.

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by Zaid-Hameed**
# Description

AWP is an important adversarial training approach because it provides better robustness against adversarial attacks and mitigates robust overfitting. AWP has been proposed in paper "Adversarial Weight Perturbation Helps
Robust Generalization".

Paper link: https://proceedings.neurips.cc/paper/2020/file/1ef91c212e30e14bf125e9374262401f-Paper.pdf

It is also a base component of more advanced adversarial training approaches.

Fixes #2164

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] Adversarial weight perturbation based training implementation produces results similar to original implementation
- [x] All functions in implemented code work as expected

**Test Configuration**:
- OS: Red Hat Enterprise Linux 8.7 (Ootpa)
- Python version: 3.9.12
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version: PyTorch 1.13.1+cu117

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @Zaid-Hameed Thank you very much for your pull request! I have added a few minor comments on using properties to avoid pylint warnings. What do you think?

Have you tested your code on GPUs?

File: art/defences/trainer/adversarial_trainer_awp_pytorch.py [hunk 1]
```diff
@@ -0,0 +1,486 @@
...
+                                  the target classifier.
+        """
+        import torch
+
+        logger.info("Performing adversarial training with AWP with %s protocol", self._mode)
+
+        # pylint: disable=W0212
+        if (scheduler is not None) and (
+            not isinstance(scheduler, torch.optim.lr_scheduler._LRScheduler)
+        ):  # pylint: enable=W0212
```
[the remark this answers is not in the document]
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_awp_pytorch.py:213]: Yes, I checked on internet and found this to be the solution when it is in the forms of a condition.

File: art/defences/trainer/adversarial_trainer_awp.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+This module implements adversarial training with AWP protocol.
```
[the remark this answers is not in the document]
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_awp.py]: Done.

File: art/defences/trainer/adversarial_trainer_awp_pytorch.py [hunk 3]
```diff
@@ -0,0 +1,486 @@
...
+if TYPE_CHECKING:
+    import torch
+
+logger = logging.getLogger(__name__)
+EPS = 1e-8
+
+
+class AdversarialTrainerAWPPyTorch(AdversarialTrainerAWP):
+    """
+    Class performing adversarial training following AWP protocol.
```
[the remark this answers is not in the document]
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_awp_pytorch.py]: Done.

File: art/defences/trainer/adversarial_trainer_awp_pytorch.py [hunk 4]
```diff
@@ -0,0 +1,486 @@
...
+from art.defences.trainer.adversarial_trainer_awp import AdversarialTrainerAWP
+from art.estimators.classification.pytorch import PyTorchClassifier
+from art.data_generators import DataGenerator
+from art.attacks.attack import EvasionAttack
+
+if TYPE_CHECKING:
+    import torch
+
+logger = logging.getLogger(__name__)
+EPS = 1e-8
```
[the remark this answers is not in the document]
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_awp_pytorch.py]: Addes definition.

File: art/defences/trainer/adversarial_trainer_awp_pytorch.py [hunk 5]
```diff
@@ -0,0 +1,486 @@
...
+        """
+        super().__init__(classifier, proxy_classifier, attack, mode, gamma, beta, warmup)
+        self._classifier: PyTorchClassifier
+        self._proxy_classifier: PyTorchClassifier
+        self._attack: EvasionAttack
+        self._mode: str
+        self.gamma: float
+        self._beta: float
+        self._warmup: int
+        self._apply_wp: bool
```
[the remark this answers is not in the document]
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_awp_pytorch.py:83]: Yes, for static type checking by mypy.

File: art/defences/trainer/adversarial_trainer_awp_pytorch.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_awp_pytorch.py]: Fixed I according to our discussion.

File: art/defences/trainer/adversarial_trainer_awp_pytorch.py [hunk 6]
```diff
@@ -0,0 +1,486 @@
...
+
+            if i_epoch >= self._warmup:
+                self._apply_wp = True
+
+            start_time = time.time()
+            train_loss = 0.0
+            train_acc = 0.0
+            train_n = 0.0
+
+            for batch_id in range(nb_batches):  # pylint: disable=W0612
```
[the remark this answers is not in the document]
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_awp_pytorch.py]: Done.

File: art/defences/trainer/adversarial_trainer_awp_pytorch.py [hunk 7]
```diff
@@ -0,0 +1,486 @@
...
+
+        if self._classifier._optimizer is None:  # pylint: disable=W0212
+            raise ValueError("Optimizer of classifier is currently None, but is required for adversarial training.")
+
+        if self._proxy_classifier._optimizer is None:  # pylint: disable=W0212
+            raise ValueError(
+                "Optimizer of proxy classifier is currently None, but is required for adversarial training."
+            )
+
+        self._classifier._model.train(mode=False)  # pylint: disable=W0212
```
[the remark this answers is not in the document]
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_awp_pytorch.py]: Done.

File: art/defences/trainer/adversarial_trainer_awp_pytorch.py [hunk 8]
```diff
@@ -0,0 +1,486 @@
...
+
+        :param x_batch: batch of x.
+        :param y_batch: batch of y.
+        :return: tuple containing batch data loss, batch data accuracy and number of samples in the batch
+        """
+        import torch
+        from torch import nn
+        import torch.nn.functional as F
+
+        if self._classifier._optimizer is None:  # pylint: disable=W0212
```
[the remark this answers is not in the document]
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_awp_pytorch.py]: Done.

File: art/defences/trainer/adversarial_trainer_awp_pytorch.py [hunk 9]
```diff
@@ -0,0 +1,486 @@
...
+            raise ValueError("Optimizer of classifier is currently None, but is required for adversarial training.")
+
+        if self._proxy_classifier._optimizer is None:  # pylint: disable=W0212
```
[the remark this answers is not in the document]
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_awp_pytorch.py]: Done.

File: art/defences/trainer/adversarial_trainer_awp_pytorch.py [hunk 10]
```diff
@@ -0,0 +1,486 @@
...
+            x_batch_pert, y_batch, fit=True
+        )
+
+        # Check label shape
+        if self._classifier._reduce_labels:  # pylint: disable=W0212
+            y_preprocessed = np.argmax(y_preprocessed, axis=1)
+
+        i_batch = torch.from_numpy(x_preprocessed).to(self._classifier._device)  # pylint: disable=W0212
+        i_batch_pert = torch.from_numpy(x_preprocessed_pert).to(self._classifier._device)  # pylint: disable=W0212
+        o_batch = torch.from_numpy(y_preprocessed).to(self._classifier._device)  # pylint: disable=W0212
```
[the remark this answers is not in the document]
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_awp_pytorch.py]: Done.

File: art/defences/trainer/adversarial_trainer_awp_pytorch.py [hunk 11]
```diff
@@ -0,0 +1,486 @@
...
+
+        self._classifier._model.train(mode=True)  # pylint: disable=W0212
```
[the remark this answers is not in the document]
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_awp_pytorch.py]: Done.

File: art/defences/trainer/adversarial_trainer_awp_pytorch.py [hunk 12]
```diff
@@ -0,0 +1,486 @@
...
+
+        if self._apply_wp:
+            w_perturb = self._weight_perturbation(x_batch=i_batch, x_batch_pert=i_batch_pert, y_batch=o_batch)
+            list_keys = list(w_perturb.keys())
+            self._modify_classifier(self._classifier, list_keys, w_perturb, op="add")
+
+        # Zero the parameter gradients
+        self._classifier._optimizer.zero_grad()  # pylint: disable=W0212
```
[the remark this answers is not in the document]
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_awp_pytorch.py]: Done.

File: art/defences/trainer/adversarial_trainer_awp_pytorch.py [hunk 13]
```diff
@@ -0,0 +1,486 @@
...
+
+        if self._mode.lower() == "pgd":
+            # Perform prediction
+            model_outputs_pert = self._classifier._model(i_batch_pert)  # pylint: disable=W0212
+            loss = self._classifier._loss(model_outputs_pert[-1], o_batch)  # pylint: disable=W0212
```
[the remark this answers is not in the document]
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_awp_pytorch.py]: Done.

File: art/defences/trainer/adversarial_trainer_awp_pytorch.py [hunk 14]
```diff
@@ -0,0 +1,486 @@
...
+
+        elif self._mode.lower() == "trades":
+            n = x_batch.shape[0]
+            # Perform prediction
+            model_outputs = self._classifier._model(i_batch)  # pylint: disable=W0212
+            model_outputs_pert = self._classifier._model(i_batch_pert)  # pylint: disable=W0212
```
[the remark this answers is not in the document]
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_awp_pytorch.py]: Done.

File: art/defences/trainer/adversarial_trainer_awp_pytorch.py [hunk 15]
```diff
@@ -0,0 +1,486 @@
...
+
+            # Form the loss function
+            loss_clean = self._classifier._loss(model_outputs[-1], o_batch)  # pylint: disable=W0212
```
[the remark this answers is not in the document]
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_awp_pytorch.py]: Done.

File: art/defences/trainer/adversarial_trainer_awp_pytorch.py [hunk 16]
```diff
@@ -0,0 +1,486 @@
...
+            loss = loss_clean + self._beta * loss_kl
+
+        else:
+            raise ValueError(
+                "Incorrect mode provided for base adversarial training. 'mode' must be among 'PGD' and 'TRADES'."
+            )
+
+        loss.backward()
+
+        self._classifier._optimizer.step()  # pylint: disable=W0212
```
[the remark this answers is not in the document]
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_awp_pytorch.py]: Done.

File: art/defences/trainer/adversarial_trainer_awp_pytorch.py [hunk 17]
```diff
@@ -0,0 +1,486 @@
...
+        w_perturb = OrderedDict()
+        params_dict, _ = self._calculate_model_params(self._classifier)
+        list_keys = list(params_dict.keys())
+        self._proxy_classifier._model.load_state_dict(self._classifier._model.state_dict())  # pylint: disable=W0212
+        self._proxy_classifier._model.train(mode=True)  # pylint: disable=W0212
+
+        if self._mode.lower() == "pgd":
+            # Perform prediction
+            model_outputs_pert = self._proxy_classifier._model(x_batch_pert)  # pylint: disable=W0212
+            loss = -self._proxy_classifier._loss(model_outputs_pert[-1], y_batch)  # pylint: disable=W0212
```
[the remark this answers is not in the document]
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_awp_pytorch.py]: Done.

File: art/defences/trainer/adversarial_trainer_awp_pytorch.py [hunk 18]
```diff
@@ -0,0 +1,486 @@
...
+        elif self._mode.lower() == "trades":
+            n = x_batch.shape[0]
+            # Perform prediction
+            model_outputs = self._proxy_classifier._model(x_batch)  # pylint: disable=W0212
+            model_outputs_pert = self._proxy_classifier._model(x_batch_pert)  # pylint: disable=W0212
+            loss_clean = self._proxy_classifier._loss(model_outputs[-1], y_batch)  # pylint: disable=W0212
```
[the remark this answers is not in the document]
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_awp_pytorch.py]: Done.

File: art/defences/trainer/adversarial_trainer_awp_pytorch.py [hunk 19]
```diff
@@ -0,0 +1,486 @@
...
+            loss = -1.0 * (loss_clean + self._beta * loss_kl)
+
+        else:
+            raise ValueError(
+                "Incorrect mode provided for base adversarial training. 'mode' must be among 'PGD' and 'TRADES'."
+            )
+
+        self._proxy_classifier._optimizer.zero_grad()  # type: ignore  # pylint: disable=W0212
+        loss.backward()
+        self._proxy_classifier._optimizer.step()  # type: ignore  # pylint: disable=W0212
```
[the remark this answers is not in the document]
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_awp_pytorch.py]: Done.

File: art/defences/trainer/adversarial_trainer_awp_pytorch.py [hunk 20]
```diff
@@ -0,0 +1,486 @@
...
+
+        params_dict_proxy, _ = self._calculate_model_params(self._proxy_classifier)
+
+        for name in list_keys:
+            perturbation = params_dict_proxy[name]["param"] - params_dict[name]["param"]
+            perturbation = torch.reshape(perturbation, params_dict[name]["size"])  # type: ignore
```
[the remark this answers is not in the document]
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_awp_pytorch.py]: The type checker expects second parameters to be tuple/list of integers and not a tensor. I have fixed it by changing the  size tensor to a list.

File: art/defences/trainer/adversarial_trainer_awp_pytorch.py [hunk 21]
```diff
@@ -0,0 +1,486 @@
...
+            w_perturb[name] = scale * perturbation
+
+        return w_perturb
+
+    # pylint: disable=R0201
+    def _calculate_model_params(
+        self, p_classifier: PyTorchClassifier
+    ) -> Tuple[Dict[str, Dict[str, "torch.Tensor"]], "torch.Tensor"]:  # pylint: enable=R0201
+        """
+
```
[the remark this answers is not in the document]
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_awp_pytorch.py:456]: Added description

File: art/defences/trainer/adversarial_trainer_awp_pytorch.py [hunk 22]
```diff
@@ -0,0 +1,486 @@
...
+        :param p_classifier: model for awp protocol.
+        :return: tuple with first element a dictionary with model parameters' names as keys and a nested dictionary
+        as value. The nested dictionary contains model parameters, model parameters' size, model parameters' norms.
+        The second element of tuple denotes norm of all model parameters
+        """
+
```
[the remark this answers is not in the document]
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_awp_pytorch.py]: Fixed.

File: art/defences/trainer/adversarial_trainer_awp_pytorch.py [hunk 23]
```diff
@@ -0,0 +1,486 @@
...
+        import torch
+
+        params_dict = OrderedDict()  # type: ignore
```
[the remark this answers is not in the document]
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_awp_pytorch.py]: Added type assignment to fix type checking error.

File: art/defences/trainer/adversarial_trainer_awp_pytorch.py [hunk 24]
```diff
@@ -0,0 +1,486 @@
...
+            perturbation = torch.reshape(perturbation, params_dict[name]["size"])  # type: ignore
+            scale = params_dict[name]["norm"] / (perturbation.norm() + EPS)
+            w_perturb[name] = scale * perturbation
+
+        return w_perturb
+
+    # pylint: disable=R0201
+    def _calculate_model_params(
+        self, p_classifier: PyTorchClassifier
+    ) -> Tuple[Dict[str, Dict[str, "torch.Tensor"]], "torch.Tensor"]:  # pylint: enable=R0201
```
[the remark this answers is not in the document]
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_awp_pytorch.py]: changed to static method.

File: art/defences/trainer/adversarial_trainer_awp_pytorch.py [hunk 25]
```diff
@@ -0,0 +1,486 @@
...
+        :return: tuple with first element a dictionary with model parameters' names as keys and a nested dictionary
+        as value. The nested dictionary contains model parameters, model parameters' size, model parameters' norms.
+        The second element of tuple denotes norm of all model parameters
+        """
+
+        import torch
+
+        params_dict = OrderedDict()  # type: ignore
+        list_params = []
+        for name, param in p_classifier._model.state_dict().items():  # pylint: disable=W0212
```
[the remark this answers is not in the document]
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_awp_pytorch.py]: Done.

File: art/defences/trainer/adversarial_trainer_awp_pytorch.py [hunk 26]
```diff
@@ -0,0 +1,486 @@
...
+
+        model_all_params = torch.cat(list_params)
+        model_all_params_norm = model_all_params.norm()
+        return params_dict, model_all_params_norm
+
+    def _modify_classifier(
+        self, p_classifier: PyTorchClassifier, list_keys: List[str], w_perturb: Dict[str, "torch.Tensor"], op: str
+    ) -> None:
+        """
+
```
[the remark this answers is not in the document]
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_awp_pytorch.py:486]: added description.

File: art/defences/trainer/adversarial_trainer_awp_pytorch.py [hunk 27]
```diff
@@ -0,0 +1,486 @@
...
+        :param p_classifier: model for awp protocol.
+        :param list_keys: list of model parameters' names
+        :param w_perturb: dictionary containing model parameters' names as keys and model parameters as values
+        :param op: controls whether weight perturbation will be added or subtracted from model parameters
+        """
+
```
[the remark this answers is not in the document]
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_awp_pytorch.py]: Done.

File: art/defences/trainer/adversarial_trainer_awp_pytorch.py [hunk 28]
```diff
@@ -0,0 +1,486 @@
...
+        import torch
+
+        if op.lower() == "add":
+            c_mult = 1.0
+        elif op.lower() == "subtract":
+            c_mult = -1.0
+        else:
+            raise ValueError("Incorrect op provided for weight perturbation. 'op' must be among 'add' and 'subtract'.")
+        with torch.no_grad():
+            for name, param in p_classifier._model.named_parameters():  # pylint: disable=W0212
```
[the remark this answers is not in the document]
Comment (Zaid-Hameed) [reply art/defences/trainer/adversarial_trainer_awp_pytorch.py]: Done.

File: examples/adversarial_training_awp.py [hunk 29]
```diff
@@ -0,0 +1,240 @@
...
+    verbose=False,
+)
+
+# Step 4: Create the trainer object - AdversarialTrainerAWPPyTorch
+trainer = AdversarialTrainerAWPPyTorch(
+    classifier, proxy_classifier, attack, mode="PGD", gamma=gamma, beta=6.0, warmup=warmup
+)
+
+
+# Build a Keras image augmentation object and wrap it in ART
```
[the remark this answers is not in the document]
Comment (Zaid-Hameed) [reply examples/adversarial_training_awp.py]: Changed to correct description.

**Pull request merged**